File: src/models/game_state.ts
import { Card } from './card';
import { Hand } from './hand';
import { Deck } from './deck';
import { Meld } from './meld';

export enum GamePhase {
  Setup = 'setup',
  Dealing = 'dealing',
  Playing = 'playing',
  Finished = 'finished',
}

export enum PlayerAction {
  Draw = 'draw',       // Bốc from Nọc
  Eat = 'eat',         // Ăn from discard pile
  Chiu = 'chiu',       // Chíu (4 of a kind)
  Discard = 'discard', // Đánh a card
  Pass = 'pass',       // Không ăn (pass turn)
  DeclareU = 'declare_u', // Declare Ù
}

export interface Player {
  readonly id: number;
  readonly name: string;
  readonly isHuman: boolean;
  readonly hand: Hand;
  readonly melds: Meld[];  // Revealed melds on table
}

export interface TurnState {
  readonly currentPlayerId: number;
  readonly phase: GamePhase;
  readonly lastAction: PlayerAction | null;
  readonly lastDiscardedCard: Card | null;
  readonly drawnCard: Card | null;
  readonly canEat: boolean;
  readonly canChiu: boolean;
  readonly hasDrawnThisTurn: boolean;
  readonly consecutivePasses: number; // for detecting a stalled/drawn round (hòa)
}

export interface GameState {
  readonly players: Player[];
  readonly deck: Deck;
  readonly turn: TurnState;
  readonly roundNumber: number;
  readonly consecutiveWins: number[];
  readonly winner: number | null;
}

export function createGameState(
  playerNames: string[],
  humanPlayerIndex: number = 0
): GameState {
  const players: Player[] = playerNames.map((name, i) => ({
    id: i,
    name,
    isHuman: i === humanPlayerIndex,
    hand: { cards: [] },
    melds: [],
  }));

  return {
    players,
    deck: { cards: [], drawPile: [], discardPile: [] },
    turn: {
      currentPlayerId: 0,
      phase: GamePhase.Setup,
      lastAction: null,
      lastDiscardedCard: null,
      drawnCard: null,
      canEat: false,
      canChiu: false,
      hasDrawnThisTurn: false,
      consecutivePasses: 0,
    },
    roundNumber: 1,
    consecutiveWins: playerNames.map(() => 0),
    winner: null,
  };
}

File: src/ChanDanGian.Core/Models/GameState.cs
namespace ChanDanGian.Core.Models;

public record Player(
    int Id,
    string Name,
    Hand Hand,
    List<IMeld> Melds,
    bool IsHuman)
{
    public static Player Create(int id, string name, bool isHuman) =>
        new(id, name, new Hand([]), [], isHuman);
}

public enum GamePhase
{
    Setup,
    Dealing,
    Playing,
    Finished
}

public enum PlayerAction
{
    Draw,
    Eat,
    Chiu,
    Discard,
    Pass,
    DeclareU
}

public record TurnState(
    int CurrentPlayerId,
    GamePhase Phase,
    PlayerAction? LastAction,
    Card? LastDiscardedCard,
    Card? DrawnCard,
    bool HasDrawnThisTurn,
    bool CanEat,
    bool CanChiu,
    int ConsecutivePasses)
{
    public static TurnState Create() => new(
        0, GamePhase.Setup, null, null, null, false, false, false, 0);
}

public record GameState(
    List<Player> Players,
    Deck Deck,
    TurnState Turn,
    int RoundNumber,
    List<int> ConsecutiveWins)
{
    public static GameState Create(List<string> playerNames)
    {
        var players = playerNames.Select((name, i) =>
            Player.Create(i, name, i == 0)).ToList();
        return new GameState(
            players,
            DeckFactory.CreateDeck(),
            TurnState.Create(),
            1,
            playerNames.Select(_ => 0).ToList());
    }
}
